### PR TITLE
NPE bug fix for the Position's serverTime

### PR DESCRIPTION
**Pull request by parveenkumardeeva**
I was not aware about the initialization of serverTime and didn't found any better way than this.

Pull Request:
https://github.com/tananaev/traccar/pull/3721

### CODE REVIEW

File: src/org/traccar/model/Position.java [hunk 1]
```diff
@@ -131,6 +131,10 @@
     public static final String ALARM_TAMPERING = "tampering";
     public static final String ALARM_REMOVING = "removing";
 
+    public Position() {
+        this.serverTime = new Date();
```
Comment (tananaev) [inline src/org/traccar/model/Position.java]: You don't need `this`.

**Comment by parveenkumardeeva** [comment]
Thanks,
Sure, In future I'll take care of it definitely.

**Comment by Abyss777** [comment]
Don't this contradict with https://github.com/tananaev/traccar/blob/master/src/org/traccar/database/DataManager.java#L341
I mean internal `serverTime` will be a bit different with that goes in DB.

**Comment by tananaev** [comment]
Not a big issue, but yes, we should probably remove that from data manager.

**Comment by tananaev** [comment]
Fixed in 68b2d3ab8897114dac8cd92618cf0777f630750a. @Abyss777, do you see any issues with the change?

**Comment by Abyss777** [comment]
No, I don't see. Good code cleanup.

But I see another problem with initial code, `Date` object creation happen twice when we loading positions from DB. First one in constructor, second in `setServerTime`. It increases amount "garbage" objects and amount used memory (not much but still) that is already problem with reports.

**Comment by tananaev** [comment]
Any ideas how to fix it? Set it every time in decoder?

**Comment by Abyss777** [comment]
May be like this?
```java
    public void setServerTime(Date serverTime) {
        if (serverTime != null) {
            if (this.serverTime != null) {
                this.serverTime.setTime(serverTime.getTime());
            } else {
                this.serverTime = new Date(serverTime.getTime());
            }
        } else {
            this.serverTime = null;
        }
    }

```

**Comment by tananaev** [comment]
How about set it in getter if it's not already set?

**Comment by Abyss777** [comment]
getter called only from `filter.skipLimit`. twice for `current` and `last` at the same time.
If we return `serverTime` initialization as was (when write in DB). `Last` do not get applied `serverTime` from DB and values in `skipLimit` will be the same.

**Comment by parveenkumardeeva** [comment]
I don't know is it right or wrong but using for other framework.
And also for the Traccar and working well:

    private Date serverTime;

    public Date getServerTime() {
        return serverTime;
    }

    public void setServerTime(Date serverTime) {
        this.serverTime = serverTime;
    }

**Comment by tananaev** [comment]
@Abyss777, main place where getter is called is `QueryBuilder`. I'll try to add new constructor.

**Comment by tananaev** [comment]
I have added a new constructor and removed new `Date` object creation every time. It was originally done to fix FindBugs issues, but I think faster code is better than some warnings. Ignored those warnings. @Abyss777, please check if you see any issues.

**Comment by Abyss777** [comment]
Good. I do not see any issues.
Just to fix is in mind, `Date` is not primitive and we should handle it more carefully as pointer.

**Pull request merged**